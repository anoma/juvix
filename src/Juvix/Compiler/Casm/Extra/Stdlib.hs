module Juvix.Compiler.Casm.Extra.Stdlib where

import Data.FileEmbed qualified as FE
import Data.Text.Encoding
import Juvix.Compiler.Casm.Data.LabelInfoBuilder
import Juvix.Compiler.Casm.Language
import Juvix.Compiler.Casm.Translation.FromSource

data StdlibBuiltins = StdlibBuiltins
  { _stdlibGetRegs :: Symbol,
    _stdlibCallClosure :: Symbol,
    _stdlibExtendClosure :: Symbol,
    _stdlibGetRegsName :: Text,
    _stdlibCallClosureName :: Text,
    _stdlibExtendClosureName :: Text
  }

makeLenses ''StdlibBuiltins

addStdlibBuiltins :: (Member LabelInfoBuilder r) => Address -> Sem r (StdlibBuiltins, [Instruction])
addStdlibBuiltins addr = do
  instrs <-
    fmap (fromRight impossible) $
      runParser' addr "stdlib.casm" $
        decodeUtf8 $(FE.makeRelativeToProject "runtime/src/casm/stdlib.casm" >>= FE.embedFile)
  let _stdlibGetRegsName :: Text = "juvix_get_regs"
      _stdlibCallClosureName :: Text = "juvix_call_closure"
      _stdlibExtendClosureName :: Text = "juvix_extend_closure"
  _stdlibGetRegs <- fromJust <$> getIdent _stdlibGetRegsName
  _stdlibCallClosure <- fromJust <$> getIdent _stdlibCallClosureName
  _stdlibExtendClosure <- fromJust <$> getIdent _stdlibExtendClosureName
  return (StdlibBuiltins {..}, instrs)
