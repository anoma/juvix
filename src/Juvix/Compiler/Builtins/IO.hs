module Juvix.Compiler.Builtins.IO where

import Juvix.Compiler.Abstract.Extra
import Juvix.Compiler.Builtins.Effect
import Juvix.Prelude

registerIO :: (Member Builtins r) => AxiomDef -> Sem r ()
registerIO d = do
  unless (isSmallUniverse' (d ^. axiomType)) (error "IO should be in the small universe")
  registerBuiltin BuiltinIO (d ^. axiomName)

registerIOSequence :: (Member Builtins r) => AxiomDef -> Sem r ()
registerIOSequence d = do
  io <- getBuiltinName (getLoc d) BuiltinIO
  unless (d ^. axiomType === (io --> io --> io)) (error "IO sequence have type IO â†’ IO")
  registerBuiltin BuiltinIOSequence (d ^. axiomName)
