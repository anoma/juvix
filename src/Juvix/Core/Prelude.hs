module Juvix.Core.Prelude
  ( module Juvix.Core.Builtins,
    Info,
    Key,
    module Juvix.Prelude,
    module Juvix.Prelude.Loc,
    module Juvix.Syntax.Abstract.Name,
    Location,
  )
where

import Juvix.Core.Builtins
import Juvix.Core.Info (Info, Key)
import Juvix.Prelude
import Juvix.Prelude.Loc
import Juvix.Syntax.Abstract.Name

type Location = Interval
