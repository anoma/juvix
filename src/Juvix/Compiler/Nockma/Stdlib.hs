{-# LANGUAGE QuasiQuotes #-}

module Juvix.Compiler.Nockma.Stdlib where

import Juvix.Compiler.Nockma.Translation.FromSource.QQ
import Juvix.Prelude

stdlib :: Term Natural
stdlib =
  [nock|
[ [ [ 7
        [ 8
        [1 1 1]
        [ 1
            8
            [1 0]
            8
            [ 1
            6
            [5 [1 0] 0 60]
            [0 6]
            9
            2
            10
            [60 8 [9 342 0 31] 9 2 10 [6 0 124] 0 2]
            10
            [6 8 [9 20 0 31] 9 2 10 [6 [0 125] 0 14] 0 2]
            0
            1
            ]
            9
            2
            0
            1
        ]
        0
        1
        ]
        11
        [1.953.718.630 1 7.107.949 [0 7] 0]
        0
        1
    ]
    [ [ 7
        [ 8
            [1 0 0]
            [ 1
            6
            [5 [1 0] 0 12]
            [0 13]
            9
            2
            10
            [6 [8 [9 342 0 7] 9 2 10 [6 0 28] 0 2] 4 0 13]
            0
            1
            ]
            0
            1
        ]
        11
        [1.953.718.630 1 6.579.297 [0 7] 0]
        0
        1
        ]
        [ [ 7
            [ 8
            [1 0 0]
            [ 1
                6
                [5 [0 12] 0 13]
                [1 0]
                6
                [8 [9 343 0 7] 9 2 10 [6 [0 28] 0 29] 0 2]
                [1 0]
                1
                1
            ]
            0
            1
            ]
            11
            [1.953.718.630 1 6.648.940 [0 7] 0]
            0
            1
        ]
        [ 7
            [ 8
            [1 1 1]
            [ 1
                6
                [5 [1 0] 0 13]
                [0 0]
                8
                [1 0]
                8
                [ 1
                6
                [8 [9 343 0 31] 9 2 10 [6 [0 124] 0 125] 0 2]
                [0 6]
                9
                2
                10
                [60 8 [9 47 0 31] 9 2 10 [6 [0 124] 0 125] 0 2]
                10
                [6 4 0 6]
                0
                1
                ]
                9
                2
                0
                1
            ]
            0
            1
            ]
            11
            [1.953.718.630 1 7.760.228 [0 7] 0]
            0
            1
        ]
        [ 7
            [ 8
            [1 0]
            [ 1
                6
                [5 [1 0] 0 6]
                [0 0]
                8
                [1 0]
                8
                [1 6 [5 [0 30] 4 0 6] [0 6] 9 2 10 [6 4 0 6] 0 1]
                9
                2
                0
                1
            ]
            0
            1
            ]
            11
            [1.953.718.630 1 6.514.020 [0 7] 0]
            0
            1
        ]
        7
        [ 8
            [1 0 0]
            [ 1
            6
            [6 [5 [0 12] 0 13] [1 1] 1 0]
            [ 6
                [ 8
                [ 1
                    6
                    [5 [1 0] 0 28]
                    [1 0]
                    6
                    [ 6
                    [6 [5 [1 0] 0 29] [1 1] 1 0]
                    [ 6
                        [ 9
                        2
                        10
                        [ 14
                            [8 [9 342 0 15] 9 2 10 [6 0 60] 0 2]
                            8
                            [9 342 0 15]
                            9
                            2
                            10
                            [6 0 61]
                            0
                            2
                        ]
                        0
                        1
                        ]
                        [1 0]
                        1
                        1
                    ]
                    1
                    1
                    ]
                    [1 0]
                    1
                    1
                ]
                9
                2
                0
                1
                ]
                [1 0]
                1
                1
            ]
            1
            1
            ]
            0
            1
        ]
        11
        [1.953.718.630 1 6.845.548 [0 7] 0]
        0
        1
        ]
        7
        [8 [1 0 0] [1 6 [8 [9 84 0 7] 9 2 10 [6 [0 28] 0 29] 0 2] [1 1] 1 0] 0 1]
        11
        [1.953.718.630 1 6.845.543 [0 7] 0]
        0
        1
    ]
    [ 7
        [ 8
        [1 0 0]
        [1 6 [8 [9 343 0 7] 9 2 10 [6 [0 28] 0 29] 0 2] [1 1] 1 0]
        0
        1
        ]
        11
        [1.953.718.630 1 6.648.935 [0 7] 0]
        0
        1
    ]
    [ 7
        [ 8
        [1 1 1]
        [ 1
            6
            [5 [1 0] 0 13]
            [0 0]
            8
            [9 47 0 7]
            9
            2
            10
            [ 6
            [0 28]
            7
            [0 3]
            8
            [9 4 0 7]
            9
            2
            10
            [6 [0 29] 7 [0 3] 8 [9 170 0 7] 9 2 10 [6 [0 28] 0 29] 0 2]
            0
            2
            ]
            0
            2
        ]
        0
        1
        ]
        11
        [1.953.718.630 1 6.582.125 [0 7] 0]
        0
        1
    ]
    7
    [ 8
        [1 0 0]
        [ 1
        6
        [5 [1 0] 0 13]
        [0 12]
        9
        2
        10
        [ 6
            [8 [9 342 0 7] 9 2 10 [6 0 28] 0 2]
            8
            [9 342 0 7]
            9
            2
            10
            [6 0 29]
            0
            2
        ]
        0
        1
        ]
        0
        1
    ]
    11
    [1.953.718.630 1 6.452.595 [0 7] 0]
    0
    1
    ]
    [0 3]
    909
]
|]
