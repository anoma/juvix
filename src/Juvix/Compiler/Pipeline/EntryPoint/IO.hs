module Juvix.Compiler.Pipeline.EntryPoint.IO where

import Juvix.Compiler.Pipeline.EntryPoint
import Juvix.Compiler.Pipeline.Root
import Juvix.Data.Effect.TaggedLock
import Juvix.Prelude

defaultEntryPointIO :: forall r. (Members '[EmbedIO, TaggedLock, EmbedIO] r) => Path Abs Dir -> Maybe (Path Abs File) -> Sem r EntryPoint
defaultEntryPointIO cwd mainFile = do
  root <- findRootAndChangeDir (parent <$> mainFile) Nothing cwd
  let pkgIdFromPackageFile :: Sem r PackageId
      pkgIdFromPackageFile = (^. packageId) <$> readPackageRootIO root
  pkgId <- case root ^. rootSomeRoot . someRootType of
    GlobalStdlib -> pkgIdFromPackageFile
    GlobalPackageDescription -> pkgIdFromPackageFile
    LocalPackage -> pkgIdFromPackageFile
    GlobalPackageBase -> return packageBaseId
  return (defaultEntryPoint pkgId root mainFile)
