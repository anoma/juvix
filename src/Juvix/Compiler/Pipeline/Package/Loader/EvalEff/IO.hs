module Juvix.Compiler.Pipeline.Package.Loader.EvalEff.IO
  ( module Juvix.Compiler.Pipeline.Package.Loader.Error,
    module Juvix.Compiler.Pipeline.Package.Loader.EvalEff.IO,
  )
where

import Data.HashMap.Strict qualified as HashMap
import Juvix.Compiler.Concrete hiding (Symbol)
import Juvix.Compiler.Concrete.Translation.FromSource.TopModuleNameChecker (runTopModuleNameChecker)
import Juvix.Compiler.Core (CoreResult, coreResultModule)
import Juvix.Compiler.Core qualified as Core
import Juvix.Compiler.Core.Evaluator
import Juvix.Compiler.Core.Extra.Value
import Juvix.Compiler.Core.Language
import Juvix.Compiler.Pipeline
import Juvix.Compiler.Pipeline.Driver (evalModuleInfoCachePackageDotJuvix, processFileToStoredCore)
import Juvix.Compiler.Pipeline.Package.Loader.Error
import Juvix.Compiler.Pipeline.Package.Loader.EvalEff
import Juvix.Compiler.Pipeline.Package.Loader.PathResolver
import Juvix.Data.Effect.Git
import Juvix.Data.Effect.Process

data LoaderResource = LoaderResource
  { _loaderResourceResult :: CoreResult,
    _loaderResourcePackagePath :: Path Abs File
  }

makeLenses ''LoaderResource

runEvalFileEffIO ::
  forall r a.
  (Members '[TaggedLock, Reader EntryPoint, Files, EmbedIO, Error PackageLoaderError] r) =>
  Sem (EvalFileEff ': r) a ->
  Sem r a
runEvalFileEffIO = runProvider_ helper
  where
    helper :: forall x. Path Abs File -> Sem (EvalEff ': r) x -> Sem r x
    helper p m = do
      res <- loadPackage' p
      let loaderRes :: LoaderResource =
            LoaderResource
              { _loaderResourceResult = res,
                _loaderResourcePackagePath = p
              }
      interpret (handler loaderRes) m

    handler :: LoaderResource -> EvalEff m x -> Sem r x
    handler res = \case
      Eval' n -> toValue tab <$> evalNode n
      LookupIdentifier n ->
        ( maybe
            ( throw
                PackageLoaderError
                  { _packageLoaderErrorPath = packagePath,
                    _packageLoaderErrorCause = ErrPackageSymbolNotFound
                  }
            )
            return
            ( identifierSymbol n
                >>= (`HashMap.lookup` (tab ^. Core.identContext))
            )
        )
      AssertNodeType n ty -> assertNodeType' n ty
      where
        tab :: Core.InfoTable
        tab = Core.computeCombinedInfoTable (res ^. loaderResourceResult . coreResultModule)

        packagePath :: Path Abs File
        packagePath = res ^. loaderResourcePackagePath

        identifiers :: HashMap Symbol Core.IdentifierInfo
        identifiers = tab ^. Core.infoIdentifiers

        identifierSymbol :: Text -> Maybe Symbol
        identifierSymbol n = fst <$> find (\(_, ii) -> ii ^. Core.identifierName == n) (HashMap.toList identifiers)

        evalNode :: Node -> Sem r Node
        evalNode n = do
          n' <- doEval Nothing False packageLoc tab n
          case n' of
            Right resN -> return resN
            Left e -> do
              throw
                PackageLoaderError
                  { _packageLoaderErrorPath = packagePath,
                    _packageLoaderErrorCause =
                      ErrPackageEvaluationError
                        PackageEvaluationError
                          { _packageEvaluationErrorError = JuvixError e
                          }
                  }
          where
            packageLoc :: Interval
            packageLoc = singletonInterval (mkInitialLoc packagePath)

        assertNodeType' :: (Foldable f) => Node -> f TypeSpec -> Sem r TypeSpec
        assertNodeType' n tys = do
          evalN <- evalNode n
          case evalN of
            NCtr Constr {..} -> do
              let ci = Core.lookupTabConstructorInfo tab _constrTag
                  ii = Core.lookupTabInductiveInfo tab (ci ^. Core.constructorInductive)
                  ty = find (checkInductiveType ii) tys
              fromMaybeM err (return ty)
            _ -> err
          where
            err :: Sem r b
            err = throw PackageLoaderError {_packageLoaderErrorPath = packagePath, _packageLoaderErrorCause = ErrPackageTypeError}

            -- Check that the type of the package identifier is the expected type from the specific
            -- PackageDescription module that is provided by the PathResolver
            checkInductiveType :: Core.InductiveInfo -> TypeSpec -> Bool
            checkInductiveType ii t = checkPackageName (t ^. typeSpecName) && checkPackageLocation (t ^. typeSpecFile)
              where
                checkPackageName :: Text -> Bool
                checkPackageName typeName = ii ^. Core.inductiveName == typeName

                checkPackageLocation :: Path Abs File -> Bool
                checkPackageLocation f = case ii ^. Core.inductiveLocation of
                  Just l -> l ^. intervalFile == f
                  Nothing -> False

loadPackage' :: (Members '[TaggedLock, Reader EntryPoint, Files, EmbedIO, Error PackageLoaderError] r) => Path Abs File -> Sem r CoreResult
loadPackage' packagePath = do
  entry <- ask @EntryPoint
  mapError toPackageError
    . runConcurrent
    . ignoreLogger
    . evalInternetOffline
    . evalHighlightBuilder
    . runProcessIO
    . runFilesIO
    . evalTopNameIdGen defaultModuleId
    . mapError (JuvixError @GitProcessError)
    . runGitProcess
    . runEvalFileEffIO
    . runPackagePathResolver rootPath
    . runTopModuleNameChecker
    . runReader noMigration
    . evalModuleInfoCachePackageDotJuvix
    $ (^. pipelineResult) <$> processFileToStoredCore (packageEntryPoint entry)
  where
    toPackageError :: JuvixError -> PackageLoaderError
    toPackageError e =
      PackageLoaderError
        { _packageLoaderErrorPath = packagePath,
          _packageLoaderErrorCause = ErrPackageJuvixError (PackageJuvixError e)
        }

    rootPath :: Path Abs Dir
    rootPath = parent packagePath

    packageEntryPoint :: EntryPoint -> EntryPoint
    packageEntryPoint entry =
      entry
        { _entryPointRoot = root ^. rootRootDir,
          _entryPointResolverRoot = root ^. rootRootDir,
          _entryPointSomeRoot = root ^. rootSomeRoot,
          _entryPointBuildDir = root ^. rootBuildDir,
          _entryPointPackageId = rootPkg,
          _entryPointModulePath = Just packagePath,
          _entryPointStdin = Nothing,
          _entryPointPipeline = Just PipelineEval
        }
      where
        sroot :: SomeRoot
        sroot =
          SomeRoot
            { _someRootDir = rootPath,
              _someRootType = GlobalPackageDescription
            }

        root :: Root
        root =
          Root
            { _rootSomeRoot = sroot,
              _rootInvokeDir = rootPath,
              _rootBuildDir = DefaultBuildDir
            }

        rootPkg :: PackageId
        rootPkg =
          PackageId
            { _packageIdVersion = defaultVersion,
              _packageIdName = "Package"
            }
