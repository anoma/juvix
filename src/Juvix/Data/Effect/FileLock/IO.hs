module Juvix.Data.Effect.FileLock.IO where

import Juvix.Data.Effect.FileLock.Base
import Juvix.Prelude
import System.FileLock hiding (FileLock)

-- | Interpret `FileLock` using `System.FileLock`
runFileLockIO :: (Members '[Resource, Embed IO] r) => Sem (FileLock ': r) a -> Sem r a
runFileLockIO = interpretH $ \case
  WithFileLock' p ma -> bracket (embed $ lockFile (toFilePath p) Exclusive) (embed . unlockFile) (const (runTSimple ma))
