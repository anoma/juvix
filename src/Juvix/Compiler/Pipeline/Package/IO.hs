module Juvix.Compiler.Pipeline.Package.IO
  ( module Juvix.Compiler.Pipeline.Package.IO,
    module Juvix.Compiler.Pipeline.Package,
  )
where

import Juvix.Compiler.Pipeline.Package
import Juvix.Compiler.Pipeline.Package.Loader
import Juvix.Compiler.Pipeline.Package.Loader.Error
import Juvix.Compiler.Pipeline.Package.Loader.EvalEff.IO
import Juvix.Data.Effect.TaggedLock
import Juvix.Prelude

loadPackageFileIO :: (Members '[TaggedLock, Error JuvixError, EmbedIO] r) => Path Abs Dir -> BuildDir -> Sem r Package
loadPackageFileIO root buildDir =
  runFilesIO
    . mapError (JuvixError @PackageLoaderError)
    . runEvalFileEffIO
    $ loadPackage buildDir (mkPackagePath root)

readPackageIO :: (Members '[TaggedLock, EmbedIO] r) => Path Abs Dir -> BuildDir -> Sem r Package
readPackageIO root buildDir =
  runFilesIO
    . runErrorIO' @JuvixError
    . mapError (JuvixError @PackageLoaderError)
    . runEvalFileEffIO
    $ readPackage root buildDir

readGlobalPackageIO :: (Members '[EmbedIO, TaggedLock] r) => Sem r Package
readGlobalPackageIO =
  runFilesIO
    . runErrorIO' @JuvixError
    . mapError (JuvixError @PackageLoaderError)
    . runEvalFileEffIO
    $ readGlobalPackage
