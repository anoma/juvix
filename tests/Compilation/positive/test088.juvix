-- Record update pun
module test088;

import Stdlib.Prelude open;
import Stdlib.System.IO open;

type R :=
  mkR@{
    a : Nat;
    b : Nat;
    c : Nat;
  };

open R using {mkR} public;

instance
RShowI : Show R :=
  Show.mk@{
    show (r : R) : String :=
      Show.show (R.a r)
        ++str " "
        ++str Show.show (R.b r)
        ++str " "
        ++str Show.show (R.c r);
  };

main : IO :=
  let
    z :=
      mkR@{
        a := 0;
        b := 0;
        c := 0;
      };
    a := 6;
    b := 7;
    c := 8;
  in printLn z >>> printLn z@R{a} >>> printLn z@R{c} >>> printLn z@R{a; b; c};
