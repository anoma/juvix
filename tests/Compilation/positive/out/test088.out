0 0 0
6 0 0
0 0 8
6 7 8
